File: agentsdk/envelope.py
"""Envelope — the core JSONL protocol wrapper.

Each envelope type uses a classmethod constructor that only sets allowed fields.
Serialization via to_json() excludes None-valued strings and zero-valued integers,
matching the Go implementation's omitempty semantics.
"""

from __future__ import annotations

import json
from dataclasses import asdict, dataclass, field
from datetime import datetime, timezone
from typing import Any, Optional

# Protocol version shared by all envelopes.
ENVELOPE_VERSION = "1.0"

# Envelope type constants.
TYPE_RESULT = "result"
TYPE_ERROR = "error"
TYPE_WARNING = "warning"
TYPE_PROGRESS = "progress"


def _utc_now_rfc3339() -> str:
    """Return the current UTC time as an RFC3339 string."""
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")


@dataclass
class Envelope:
    """Top-level JSONL protocol envelope.

    Field exclusion is enforced via constructors — only the fields appropriate
    for each type are set. ``to_json()`` serializes, omitting None / 0 values
    to match the Go ``omitempty`` behaviour.
    """

    version: str = ENVELOPE_VERSION
    tool: str = ""
    type: str = ""
    timestamp: str = field(default_factory=_utc_now_rfc3339)
    data: Optional[Any] = None
    error_code: Optional[str] = None
    message: Optional[str] = None
    percent: int = 0
    trace_id: Optional[str] = None

    # ------------------------------------------------------------------
    # Constructors (one per envelope type)
    # ------------------------------------------------------------------

    @classmethod
    def result(cls, *, tool: str, data: Any) -> Envelope:
        """Create a result envelope — carries ``data`` only."""
        return cls(tool=tool, type=TYPE_RESULT, data=data)

    @classmethod
    def error(cls, *, tool: str, error_code: str, message: str) -> Envelope:
        """Create an error envelope — carries ``error_code`` + ``message``."""
        return cls(tool=tool, type=TYPE_ERROR, error_code=error_code, message=message)

    @classmethod
    def warning(cls, *, tool: str, message: str) -> Envelope:
        """Create a warning envelope — carries ``message`` only."""
        return cls(tool=tool, type=TYPE_WARNING, message=message)

    @classmethod
    def progress(cls, *, tool: str, percent: int, message: str) -> Envelope:
        """Create a progress envelope — carries ``percent`` + ``message``."""
        return cls(tool=tool, type=TYPE_PROGRESS, percent=percent, message=message)

    # ------------------------------------------------------------------
    # Serialization
    # ------------------------------------------------------------------

    def to_dict(self) -> dict[str, Any]:
        """Convert to a dict, excluding None strings and zero ints.

        This mirrors Go's ``omitempty`` JSON tag semantics:
        - ``None`` string/Any fields → omitted
        - ``0`` int fields (percent) → omitted
        - ``""`` empty string fields (version/tool/type/timestamp) are kept
          because they are structurally required.
        """
        d: dict[str, Any] = {}
        for key, value in asdict(self).items():
            if value is None:
                continue
            if isinstance(value, int) and value == 0:
                continue
            d[key] = value
        return d

    def to_json(self) -> str:
        """Serialize to a JSON string, excluding zero-valued fields."""
        return json.dumps(self.to_dict(), separators=(",", ":"))

File: sdks/go/envelope.go
package agentsdk

import "time"

// EnvelopeVersion is the protocol version for all envelopes.
const EnvelopeVersion = "1.0"

// Envelope type constants.
const (
	TypeResult   = "result"
	TypeError    = "error"
	TypeWarning  = "warning"
	TypeProgress = "progress"
)

// Envelope is the top-level JSONL wrapper. Field exclusion is enforced via
// constructors — zero-value fields with omitempty are omitted from JSON output.
type Envelope struct {
	Version   string      `json:"version"`
	Tool      string      `json:"tool"`
	Type      string      `json:"type"`
	Timestamp string      `json:"timestamp"`
	Data      interface{} `json:"data,omitempty"`
	ErrorCode string      `json:"error_code,omitempty"`
	Message   string      `json:"message,omitempty"`
	Percent   int         `json:"percent,omitempty"`
	TraceID   string      `json:"trace_id,omitempty"`
	Kind      string      `json:"kind,omitempty"`
}

// NewResultEnvelope creates a result envelope with data only.
// Fields error_code and percent are guaranteed zero-valued (omitted from JSON).
// When kind is non-empty, a top-level "kind" field is included in the JSON output.
func NewResultEnvelope(tool string, data interface{}, kind ...string) Envelope {
	var k string
	if len(kind) > 0 && kind[0] != "" {
		k = kind[0]
	}
	return Envelope{
		Version:   EnvelopeVersion,
		Tool:      tool,
		Type:      TypeResult,
		Timestamp: time.Now().UTC().Format(time.RFC3339),
		Data:      data,
		Kind:      k,
	}
	return Envelope{
		Version:   EnvelopeVersion,
		Tool:      tool,
		Type:      TypeResult,
		Timestamp: time.Now().UTC().Format(time.RFC3339),
		Data:      data,
	}
}

// NewErrorEnvelope creates an error envelope with error_code and message only.
// Fields data and percent are guaranteed zero-valued (omitted from JSON).
func NewErrorEnvelope(tool string, errorCode string, message string) Envelope {
	return Envelope{
		Version:   EnvelopeVersion,
		Tool:      tool,
		Type:      TypeError,
		Timestamp: time.Now().UTC().Format(time.RFC3339),
		ErrorCode: errorCode,
		Message:   message,
	}
}

// NewWarningEnvelope creates a warning envelope with message only.
// Fields data, error_code and percent are guaranteed zero-valued (omitted from JSON).
func NewWarningEnvelope(tool string, message string) Envelope {
	return Envelope{
		Version:   EnvelopeVersion,
		Tool:      tool,
		Type:      TypeWarning,
		Timestamp: time.Now().UTC().Format(time.RFC3339),
		Message:   message,
	}
}

// NewProgressEnvelope creates a progress envelope with percent and optional message.
// Fields data and error_code are guaranteed zero-valued (omitted from JSON).
func NewProgressEnvelope(tool string, percent int, message string) Envelope {
	return Envelope{
		Version:   EnvelopeVersion,
		Tool:      tool,
		Type:      TypeProgress,
		Timestamp: time.Now().UTC().Format(time.RFC3339),
		Percent:   percent,
		Message:   message,
	}
}
